compiler/src/dotty/tools/dotc/util/LinearSet.scala

```diff
@@ -0,0 +1,45 @@
+package dotty.tools.dotc.util
+import collection.immutable
+
+/** A linear set is a set here after a `+` the previous set value cannot be
+ *  used anymore. The set is implemented as an immutable set for sizes <= 4
+ *  and as a HashSet for larger sizes.
+ */
+opaque type LinearSet[Elem >: Null <: AnyRef] =
+  immutable.Set[Elem] | HashSet[Elem]
```

Comment:
Minor: `LinearSet` is not a great name in the context of data structures, `linear` associates closer to complexity theory than type theory :)

Reply:
What would be a good other name?